src/cg.jl

```diff
@@ -47,18 +57,19 @@ function iterate(it::CGIterable, iteration::Int=start(it))
     end
 
     # u := r + βu (almost an axpy)
-    β = it.residual^2 / it.prev_residual^2
+    ρ = _dot(it.r, it.conjugate_dot)
```

Comment:
CG and preconditioned CG are split into separate methods because CG requires one fewer dot product. Can you check if you can drop this dot in your product somehow?

Reply:
Good point, but I'm afraid we can't.  It turns out that the saving of one dot product in the non-preconditioned CG comes from the fact that we calculate `ρ = r⋅r` and the norm of the residual vector can be calculated by `sqrt(ρ)` instead of performing the dot product again.

In COCG, we use the unconjugated dot product, so `ρ = transpose(r)*r`.  Therefore, the norm of the residual vector cannot be calculated by `sqrt(ρ)`, but needs to be calculated by `sqrt(r⋅r)`.

Reply:
In that case you should probably have
```jl
β = it.conjugate_dot isa Val{true} ? it.residual^2 / it.prev_residual^2 : _dot(it.r, it.conjugate_dot)
```
so that it only does the extra dot product in the unconjugated case.

Reply:
And if you do that please change `Val{true}` into a named type to improve the readability a bit

Reply:
Done!